sonaranalyzer-dotnet/src/SonarAnalyzer.Common/Helpers/SonarAnalysisContext.cs

```diff
@@ -145,57 +152,57 @@ internal void RegisterCodeBlockStartAction<TLanguageKindEnum>(Action<CodeBlockSt
         internal void RegisterSyntaxNodeAction<TLanguageKindEnum>(Action<SyntaxNodeAnalysisContext> action,
             params TLanguageKindEnum[] syntaxKinds)
             where TLanguageKindEnum : struct =>
-            context.RegisterSyntaxNodeAction(
-                c =>
-                {
-                    if (!IsAnalysisDisabled(c.GetSyntaxTree(), supportedDiagnostics))
+            RegisterContextAction(() =>
+                context.RegisterSyntaxNodeAction(
+                    c =>
                     {
-                        action(c);
-                    }
-                }, syntaxKinds);
+                        if (!IsAnalysisDisabled(c.GetSyntaxTree()))
+                        {
+                            action(c);
+                        }
+                    }, syntaxKinds));
 
         internal void RegisterSyntaxTreeAction(Action<SyntaxTreeAnalysisContext> action) =>
-            context.RegisterSyntaxTreeAction(
-                c =>
-                {
-                    if (!IsAnalysisDisabled(c.GetSyntaxTree(), supportedDiagnostics))
+            RegisterContextAction(() =>
+                context.RegisterSyntaxTreeAction(
+                    c =>
                     {
-                        action(c);
-                    }
-                });
+                        if (!IsAnalysisDisabled(c.GetSyntaxTree()))
+                        {
+                            action(c);
+                        }
+                    }));
 
         internal void RegisterSymbolAction(Action<SymbolAnalysisContext> action, ImmutableArray<SymbolKind> symbolKinds) =>
             RegisterSymbolAction(action, symbolKinds.ToArray());
 
         public void RegisterSymbolAction(Action<SymbolAnalysisContext> action, params SymbolKind[] symbolKinds) =>
-            context.RegisterSymbolAction(
-                c =>
-                {
-                    if (!IsAnalysisDisabled(c.GetSyntaxTree(), supportedDiagnostics))
+            RegisterContextAction(() =>
+                context.RegisterSymbolAction(
+                    c =>
                     {
-                        action(c);
-                    }
-                }, symbolKinds);
+                        if (!IsAnalysisDisabled(c.GetSyntaxTree()))
+                        {
+                            action(c);
+                        }
+                    }, symbolKinds));
 
-        internal static bool IsAnalysisDisabled(SyntaxTree tree, IEnumerable<DiagnosticDescriptor> supportedDiagnostics = null)
+        private void RegisterContextAction(Action action)

```

Comment:
Please suggest a better name!

Reply:
I can't think of one!

Reply:
It would be worth adding a comment somewhere to say why we have such a complicated registration/suppression mechanism. The following is a pretty lengthy explanation, but it will help us with future maintenance:

For performance reasons, we don't want to register callbacks for rules that should not be run. However, we don't necessarily have enough information to make the decision up front, so the check is broken into two parts:
1) pre-registration: at this point we only know the rules that are being registered. If we can decide from the diagnostic descriptors that a rule should not be run then we won't register the callback with the analysis context.
2) post-registration: we are now being called back by the analysis context, so we now have access to the syntax tree (and project) that are being analysed, and can use that information to decide whether to execute the logic of the rule.